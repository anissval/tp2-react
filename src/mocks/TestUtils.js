// test-utils.js
import React from 'react'
import {render as rtlRender} from '@testing-library/react'
import {createStore} from 'redux'
import {Provider} from 'react-redux'

// my reducer
import {heroesListManager} from "../reducer";
import {BrowserRouter} from "react-router-dom";

const renderWithRouterAndRedux = (ui, {
    route = '/',
    initialState,
    store = createStore(heroesListManager, initialState),
    ...renderOptions
} = {}) => {
    window.history.pushState({}, 'Test page', route);

    const ReduxWrapper = ({children}) => {
        return <Provider store={store}><BrowserRouter>{children}</BrowserRouter></Provider>
    }
    return rtlRender(ui, {wrapper: ReduxWrapper, ...renderOptions});
};

const renderWithRedux = (ui, {
    initialState,
    store = createStore(heroesListManager, initialState),
    ...renderOptions
} = {}) => {
    const ReduxWrapper = ({children}) => {
        return <Provider store={store}>{children}</Provider>
    }
    return rtlRender(ui, {wrapper: ReduxWrapper, ...renderOptions});
};
// re-export everything
export * from '@testing-library/react'
// override render method
export {renderWithRouterAndRedux, renderWithRedux};
