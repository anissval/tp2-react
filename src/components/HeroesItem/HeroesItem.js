import "../heroes/Heroes.css"
import {useDispatch, useSelector} from "react-redux";
import {removeHero} from "../../actions";
import {Link} from "react-router-dom";

export const HeroesItem = (props) => {
    const heroList = useSelector(state => state.heroesList);
    const dispatch = useDispatch();

    const listItems = heroList.map((hero) => (
        <li key={`heroItem-${hero.id}`}>
            <Link to={{
                pathname: `/heroDetails/${hero.id}`
            }}>< span className="badge">{hero.id}</span> {hero.name}</Link>
            <button className="delete" title="delete hero" onClick={() => dispatch(removeHero(hero.id))} data-testid={`delete-button-${hero.id}`}>x</button>
        </li>
    ));
    return (
        <ul className="heroes">
            {listItems}
        </ul>
    );
}
