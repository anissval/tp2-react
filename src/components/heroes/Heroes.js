import React, {useState} from "react";
import {useDispatch} from "react-redux";
import {HeroesItem} from "../HeroesItem/HeroesItem";
import {addHero} from "../../actions";
import {useHistory} from "react-router-dom";
import "./Heroes.css";

export const Heroes = (props) => {

    const dispatch = useDispatch();
    const  [heroName, setHeroName] = useState("");
    const history = useHistory();
    function handleOnChange(e) {
        e.preventDefault();
        setHeroName(e.target.value);
    }

    function onSubmitForm () {
        dispatch(addHero(heroName));
        setHeroName("")
        history.push('/');
    }
        return (
            <div className="todoListMain">
                <div className="header">
                    <label id="new-hero">Hero name: </label>
                    <form onSubmit={onSubmitForm}>
                        <input placeholder="enter hero" value={heroName} onChange={handleOnChange} data-testid="add-input">
                        </input>
                        <button className="add-button" type={"submit"} data-testid="add-button">add</button>
                    </form>
                </div>
                <HeroesItem/>
            </div>
        );
}
