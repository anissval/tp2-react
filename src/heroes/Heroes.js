import React, {useState} from "react";
import {useDispatch} from "react-redux";
import {HeroesItem} from "./HeroesItem";
import {addHero} from "../actions";
import {useHistory} from "react-router-dom";
import "./Heroes.css";

export const Heroes = (props) => {

    const dispatch = useDispatch();
    const  [heroName, setHeroName] = useState("");
    const history = useHistory();
    function handleOnChange(e) {
        e.preventDefault();
        setHeroName(e.target.value);
    }

    function onSubmitForm () {
        dispatch(addHero(heroName));
        setHeroName("")
        history.push('/');
    }
        return (
            <div className="todoListMain">
                <div className="header">
                    <label id="new-hero">Hero name: </label>
                    <form onSubmit={onSubmitForm}>
                        <input placeholder="enter hero" value={heroName} onChange={handleOnChange}>
                        </input>
                        <button className="add-button" type={"submit"}>add</button>
                    </form>
                </div>
                <HeroesItem/>
            </div>
        );
}
