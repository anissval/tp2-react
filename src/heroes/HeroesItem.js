import {Component} from "react";
import "./Heroes.css"
import {connect} from "react-redux";

var addHeroAction = {type: "add"};
var removeHeroAction = {type: "remove"};

class HeroesItem extends Component {
    constructor(props) {
        super(props);
    }

    delete(id) {
        this.props.delete(id);
    }

    render() {
        var listItems = this.props.heroesList.map((hero) =>
            <div>
                <li key={hero.id}>
                    <a>
                        <span className="badge">{hero.id}</span> {hero.name}
                    </a>
                    <button className="delete" onClick={() => this.props.removeHero(hero.id)}>x</button>
                </li>
            </div>
        );
        return (
            <ul class="heroes">
                {listItems}
            </ul>
        );
    }

}

function mapStateToProps(state) {
    return { heroesList : state.heroesList }
}

function mapDispatchToProps(dispatch) {
    return {
        addHero: function () {
            return dispatch(addHeroAction);
        },
        removeHero: function () {
            return dispatch(removeHeroAction);
        }
    }
}

export default connect(mapStateToProps, mapDispatchToProps) (HeroesItem);
