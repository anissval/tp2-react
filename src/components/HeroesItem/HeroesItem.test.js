import React from 'react';
import {HeroesItem} from "./HeroesItem";
import {fireEvent, getByTestId, prettyDOM, render, screen} from "@testing-library/react";
import {renderWithRouterAndRedux} from "../../mocks/TestUtils";
import heroesMock from "../../mocks/HeroesMock";
import {Heroes} from "../heroes/Heroes";

const hero = heroesMock[1];

describe('HeroesItem', () => {
    let component;
    beforeEach(() => {
        component = <HeroesItem/>;
    });
    test('renders Heroes Item component', () => {
        const tree = renderWithRouterAndRedux(component, '/dashboard');
        console.log(prettyDOM(tree.container));
        expect(tree).toMatchSnapshot();
    });

    test("delete a Hero", () => {
        const tree = renderWithRouterAndRedux(component, '/dashboard');
        console.log(prettyDOM(tree.container));
        const heroName = screen.queryByText(hero.name);
        expect(heroName).toBeInTheDocument();
        const button = getByTestId(tree.container, `delete-button-${hero.id}`);
        fireEvent.click(button);
        expect(heroName).not.toBeInTheDocument();
    });
});
