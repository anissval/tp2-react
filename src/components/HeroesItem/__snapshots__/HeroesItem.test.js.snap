// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeroesItem renders Heroes Item component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <ul
        class="heroes"
      >
        <li>
          <a
            href="/heroDetails/13"
          >
            <span
              class="badge"
            >
              13
            </span>
             
            Bombasto
          </a>
          <button
            class="delete"
            data-testid="delete-button-13"
            title="delete hero"
          >
            x
          </button>
        </li>
        <li>
          <a
            href="/heroDetails/14"
          >
            <span
              class="badge"
            >
              14
            </span>
             
            Celeritas
          </a>
          <button
            class="delete"
            data-testid="delete-button-14"
            title="delete hero"
          >
            x
          </button>
        </li>
        <li>
          <a
            href="/heroDetails/15"
          >
            <span
              class="badge"
            >
              15
            </span>
             
            Magneta
          </a>
          <button
            class="delete"
            data-testid="delete-button-15"
            title="delete hero"
          >
            x
          </button>
        </li>
        <li>
          <a
            href="/heroDetails/16"
          >
            <span
              class="badge"
            >
              16
            </span>
             
            RubberMan
          </a>
          <button
            class="delete"
            data-testid="delete-button-16"
            title="delete hero"
          >
            x
          </button>
        </li>
        <li>
          <a
            href="/heroDetails/17"
          >
            <span
              class="badge"
            >
              17
            </span>
             
            Dynama
          </a>
          <button
            class="delete"
            data-testid="delete-button-17"
            title="delete hero"
          >
            x
          </button>
        </li>
        <li>
          <a
            href="/heroDetails/18"
          >
            <span
              class="badge"
            >
              18
            </span>
             
            Dr IQ
          </a>
          <button
            class="delete"
            data-testid="delete-button-18"
            title="delete hero"
          >
            x
          </button>
        </li>
        <li>
          <a
            href="/heroDetails/19"
          >
            <span
              class="badge"
            >
              19
            </span>
             
            Magma
          </a>
          <button
            class="delete"
            data-testid="delete-button-19"
            title="delete hero"
          >
            x
          </button>
        </li>
        <li>
          <a
            href="/heroDetails/20"
          >
            <span
              class="badge"
            >
              20
            </span>
             
            Tornado
          </a>
          <button
            class="delete"
            data-testid="delete-button-20"
            title="delete hero"
          >
            x
          </button>
        </li>
      </ul>
    </div>
  </body>,
  "container": <div>
    <ul
      class="heroes"
    >
      <li>
        <a
          href="/heroDetails/13"
        >
          <span
            class="badge"
          >
            13
          </span>
           
          Bombasto
        </a>
        <button
          class="delete"
          data-testid="delete-button-13"
          title="delete hero"
        >
          x
        </button>
      </li>
      <li>
        <a
          href="/heroDetails/14"
        >
          <span
            class="badge"
          >
            14
          </span>
           
          Celeritas
        </a>
        <button
          class="delete"
          data-testid="delete-button-14"
          title="delete hero"
        >
          x
        </button>
      </li>
      <li>
        <a
          href="/heroDetails/15"
        >
          <span
            class="badge"
          >
            15
          </span>
           
          Magneta
        </a>
        <button
          class="delete"
          data-testid="delete-button-15"
          title="delete hero"
        >
          x
        </button>
      </li>
      <li>
        <a
          href="/heroDetails/16"
        >
          <span
            class="badge"
          >
            16
          </span>
           
          RubberMan
        </a>
        <button
          class="delete"
          data-testid="delete-button-16"
          title="delete hero"
        >
          x
        </button>
      </li>
      <li>
        <a
          href="/heroDetails/17"
        >
          <span
            class="badge"
          >
            17
          </span>
           
          Dynama
        </a>
        <button
          class="delete"
          data-testid="delete-button-17"
          title="delete hero"
        >
          x
        </button>
      </li>
      <li>
        <a
          href="/heroDetails/18"
        >
          <span
            class="badge"
          >
            18
          </span>
           
          Dr IQ
        </a>
        <button
          class="delete"
          data-testid="delete-button-18"
          title="delete hero"
        >
          x
        </button>
      </li>
      <li>
        <a
          href="/heroDetails/19"
        >
          <span
            class="badge"
          >
            19
          </span>
           
          Magma
        </a>
        <button
          class="delete"
          data-testid="delete-button-19"
          title="delete hero"
        >
          x
        </button>
      </li>
      <li>
        <a
          href="/heroDetails/20"
        >
          <span
            class="badge"
          >
            20
          </span>
           
          Tornado
        </a>
        <button
          class="delete"
          data-testid="delete-button-20"
          title="delete hero"
        >
          x
        </button>
      </li>
    </ul>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
